nexus/src/db/model.rs

```diff
@@ -1291,22 +1340,59 @@ pub struct VpcSubnet {
     identity: VpcSubnetIdentity,
 
     pub vpc_id: Uuid,
-    pub ipv4_block: Option<Ipv4Net>,
-    pub ipv6_block: Option<Ipv6Net>,
+    pub ipv4_block: Ipv4Net,
+    pub ipv6_block: Ipv6Net,
 }
 
 impl VpcSubnet {
+    /// Create a new VPC Subnet.
+    ///
+    /// NOTE: This assumes that the IP address ranges provided in `params` are
+    /// valid for the VPC, and does not do any further validation.
     pub fn new(
         subnet_id: Uuid,
         vpc_id: Uuid,
-        params: params::VpcSubnetCreate,
+        identity: external::IdentityMetadataCreateParams,
```

Comment:
It seems like we're changing the calling convention from the one that's used elsewhere in this file (namely, accepting the `params::*Create` version of this struct in `new`).  For comparison, I checked `Zpool`, `Organization`, `Project`, `Instance`, `Disk`, `OximeterInfo`, `VpcSubnet`, `VpcRouter`.  There are a few that _don't_ look this way (e.g., `RouterRoute`, `RouteDestination`, `VpcFirewallRule`, and some others) but I think those are the exception and I'm not sure all of them have an analogous `Create` struct.  Anyway, it seems like there's a convention here and we're undoing it -- is there a reason for that?

Reply:
Another good question. I think there was a bit of evolution as well as design here, but here's the reasoning I _think_ I had. Basically, we usually need to take the `params::VpcSubnetCreate` and modify it or at least check the values it contains _before_ creating the actual `VpcSubnet` from it. Most of the other methods that accept the corresponding type in the `params` module don't do any validation, but pass the values on unchanged.

For example, the params type has the IPv6 range as an `Option`. If it's not provided, we need to create it based on the IPv6 prefix of the VPC. We can do that, replacing the option in that case, and then pass a known-good value in the `Some` variant. But in that case the `Option` seems to be a red herring -- there's nothing optional about it, and at this point it's already been verified, and always going to be `Some`.

There are alternatives. We could pass a reference to the `Vpc` object, rather than just its UUID, and do the validation on the address (or random subnet generation) inside the `VpcSubnet::new` method. That would then return a `Result`, just like some of the others. There are probably other approaches, so I'm happy to explore those too!